test/samplers/guaranteed_throughput_sampler.js

```diff
@@ -69,53 +69,76 @@ describe('GuaranteedThroughput sampler', () => {
         sampler.close();
     });
 
-    let assertValues = function assertValues(sampler, lb, rate) {
-        assert.equal(lb, sampler._lowerBoundSampler.maxTracesPerSecond);
+    it('should rate limited by upper bound rate limiter', () => {
+        let sampler = new GuaranteedThroughputSampler(1, 1, 2);
+
+        let expectedTags = {'sampler.type': 'probabilistic', 'sampler.param': 1};
+        [1, 2, 3].forEach((_) => {
+            let actualTags = {};
+            assert.isOk(sampler.isSampled('testOperationName', actualTags), 'must sample');
+            assert.deepEqual(expectedTags, actualTags);
+        });
+        // The max_samples_per_second is 2, so after three calls, the upperBoundRateLimiter
+        // should be triggered and the samplingRate should be halved.
+        assertValues(sampler, 0.5, 1, 2);
+
+        sampler.close();
+    });
+
+    let assertValues = function assertValues(sampler, rate, lb, up) {
         assert.equal(rate, sampler._probabilisticSampler.samplingRate);
+        assert.equal(lb, sampler._lowerBoundSampler.maxTracesPerSecond);
+        assert.equal(up, sampler._upperBoundRateLimiter.creditsPerSecond);
     };
 
     it('should not change when update() called with the same values', () => {
-        let sampler = new GuaranteedThroughputSampler(2, 1.0);
-        assertValues(sampler, 2, 1.0);
+        let sampler = new GuaranteedThroughputSampler(1.0, 2, 3);
+        assertValues(sampler, 1.0, 2, 3);
 
         let p1 = sampler._probabilisticSampler;
         let p2 = sampler._lowerBoundSampler;
-        let isUpdated: boolean = sampler.update(2, 1.0);
+        let p3 = sampler._upperBoundRateLimiter;
+        let isUpdated: boolean = sampler.update(1.0, 2, 3);
         assert.isFalse(isUpdated);
         assert.strictEqual(sampler._probabilisticSampler, p1);
         assert.strictEqual(sampler._lowerBoundSampler, p2);
-        assertValues(sampler, 2, 1.0);
+        assert.strictEqual(sampler._upperBoundRateLimiter, p3);
+        assertValues(sampler, 1.0, 2, 3);
```

Comment:
same concept as above, since these assertValues match the sampler params then use the same variable.